File: cheater-pi/src/reliability/failureCompressor.ts
// Deterministic failure compressor: turn raw pytest / jest / vitest / tsc / node output into
// a tight structured card the model can act on. Raw test logs are the single worst context
// killer for a small local model; a ~10-line card with expected/got, the top stack file, and
// one imperative next-action line is what a weak instruction-follower localizes from.
//
// Ported from the legacy Python cheater/failure_compressor.py, extended to JS/TS toolchains.

export interface FailureEntry {
  test: string;
  errorType: string;
  message: string;
  expected: string;
  got: string;
  topStackFiles: string[];
}

export interface FailureSummary {
  command: string;
  passed: boolean;
  exitCode: number;
  errorType: string;
  summary: string;
  isSyntaxError: boolean;
  isImportError: boolean;
  isTestFailure: boolean;
  failures: FailureEntry[];
  nextAction: string;
  tail: string;
}

function stripAnsi(text: string): string {
  // eslint-disable-next-line no-control-regex
  return (text ?? "").replace(/\x1b\[[0-9;]*[a-zA-Z]/g, "");
}

function clip(text: string, maxChars: number): string {
  if (!text || text.length <= maxChars) return text ?? "";
  return `${text.slice(0, Math.max(0, maxChars - 15))}\n... [truncated]`;
}

function tail(text: string, maxLines = 60, maxChars = 1200): string {
  if (!text) return "";
  const lines = text.split(/\r?\n/);
  return clip(lines.slice(-maxLines).join("\n"), maxChars);
}

function uniq(items: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const item of items) {
    if (item && !seen.has(item)) { seen.add(item); out.push(item); }
  }
  return out;
}

const PY_FILE_LINE = /File ["']([^"']+)["'], line (\d+)/g;
const JS_STACK_FILE = /\(?((?:[A-Za-z]:)?[\w./\\-]+\.(?:[jt]sx?|mjs|cjs)):(\d+)(?::\d+)?\)?/g;
const TSC_LINE = /^([\w./\\-]+\.tsx?)\((\d+),\d+\):\s*error\s+(TS\d+):\s*(.+)$/m;
const GENERIC_ERR = /\b(\w+(?:Error|Exception|Warning))\b/;

function stackFiles(output: string, re: RegExp): string[] {
  const files: string[] = [];
  for (const m of output.matchAll(re)) files.push(m[1].replace(/\\/g, "/"));
  return uniq(files);
}

/** Parse raw command output into a structured failure summary. Never throws. */
export function compressCommandResult(command: string, stdout: string, stderr: string, exitCode: number): FailureSummary {
  const out = stripAnsi(`${stdout ?? ""}\n${stderr ?? ""}`);
  const summary: FailureSummary = {
    command, passed: exitCode === 0, exitCode, errorType: "", summary: "",
    isSyntaxError: false, isImportError: false, isTestFailure: false,
    failures: [], nextAction: "", tail: tail(out)
  };
  if (summary.passed) {
    summary.summary = "command exited 0";
    summary.nextAction = "No action needed; command passed.";
    return summary;
  }

  // 1) TypeScript compiler errors
  const tsc = out.match(TSC_LINE);
  if (tsc) {
    summary.isSyntaxError = /TS1\d{3}/.test(tsc[3]);
    summary.errorType = tsc[3];
    summary.summary = `${tsc[3]} at ${tsc[1]}:${tsc[2]}: ${tsc[4]}`.slice(0, 240);
    const files = uniq([...out.matchAll(/^([\w./\\-]+\.tsx?)\(\d+,\d+\):\s*error/gm)].map((m) => m[1].replace(/\\/g, "/")));
    summary.failures.push({ test: "(tsc)", errorType: tsc[3], message: tsc[4].slice(0, 240), expected: "", got: "", topStackFiles: files.slice(0, 5) });
    summary.nextAction = `Fix the type error at ${tsc[1]}:${tsc[2]} (${tsc[3]}).`;
    return summary;
  }

  // 2) Python syntax error
  const syntax = out.match(/File ["']([^"']+)["'], line (\d+)[\s\S]{0,200}?(SyntaxError|IndentationError|TabError)/);
  if (syntax) {
    summary.isSyntaxError = true;
    summary.errorType = syntax[3];
    summary.summary = `Python ${syntax[3]} at ${syntax[1]}:L${syntax[2]}`;
    summary.failures.push({ test: "(syntax)", errorType: syntax[3], message: summary.summary, expected: "", got: "", topStackFiles: [syntax[1].replace(/\\/g, "/")] });
    summary.nextAction = `Fix the ${syntax[3]} at ${syntax[1]}:${syntax[2]}.`;
    return summary;
  }

  // 3) Import / module resolution errors
  const modNotFound = out.match(/(?:ModuleNotFoundError: No module named ['"]([^'"]+)['"]|Cannot find module ['"]([^'"]+)['"])/);
  if (modNotFound || /ImportError|ERR_MODULE_NOT_FOUND/.test(out)) {
    summary.isImportError = true;
    const mod = modNotFound?.[1] ?? modNotFound?.[2] ?? "";
    summary.errorType = modNotFound ? "ModuleNotFound" : "ImportError";
    summary.summary = mod ? `Missing module: ${mod}` : "import failed";
    const files = uniq([...stackFiles(out, PY_FILE_LINE), ...stackFiles(out, JS_STACK_FILE)]);
    summary.failures.push({ test: "(import)", errorType: summary.errorType, message: summary.summary, expected: "", got: "", topStackFiles: files.slice(0, 5) });
    summary.nextAction = mod ? `Resolve the missing import "${mod}" (fix the path or install/pin the dependency).` : "Fix the failing import.";
    return summary;
  }

  // 4) pytest failures - match both the short-summary form ("FAILED path::test - msg")
  // and the per-test form ("path::test FAILED").
  const pytestFailed = uniq([
    ...[...out.matchAll(/^FAILED\s+([\w./:\\-]+?)(?:\s*-\s*|\s*$)/gm)].map((m) => m[1]),
    ...[...out.matchAll(/^([\w./\\-]+\.py(?:::[\w\[\]-]+)*)\s+FAILED\b/gm)].map((m) => m[1])
  ]);
  if (pytestFailed.length || (/\bFAILED\b/.test(out) && /\d+ failed/.test(out))) {
    summary.isTestFailure = true;
    const shortSummary = out.match(/=+\s*((?:\d+\s+)?(?:passed|failed|error|skipped)[^\n=]*?)\s*=+/);
    summary.summary = shortSummary ? shortSummary[1].trim() : `${pytestFailed.length} test(s) failed`;
    for (const testName of pytestFailed.slice(0, 6)) {
      const entry: FailureEntry = { test: testName, errorType: "", message: "", expected: "", got: "", topStackFiles: [] };
      const body = pytestFailureBody(out, testName);
      entry.topStackFiles = stackFiles(body, PY_FILE_LINE).slice(0, 5);
      if (!entry.topStackFiles.length && testName.includes("::")) entry.topStackFiles = [testName.split("::")[0]];
      entry.errorType = body.match(GENERIC_ERR)?.[1] ?? "";
      entry.message = (body.match(/AssertionError:?\s*(.*)/i)?.[1]
        ?? body.split(/\r?\n/).map((l) => l.trim()).find((l) => l.startsWith("E "))?.replace(/^E\s+/, "")
        ?? "").slice(0, 240);
      [entry.expected, entry.got] = extractExpectedGot(body);
      summary.failures.push(entry);
    }
    // Never leave a test failure with an empty failures list (the fallback path when the
    // "N failed" line fired but no test id was parseable).
    if (!summary.failures.length) {
      const [expected, got] = extractExpectedGot(out);
      summary.failures.push({ test: "(test)", errorType: out.match(GENERIC_ERR)?.[1] ?? "", message: (out.match(/AssertionError:?\s*(.*)/i)?.[1] ?? "").slice(0, 240), expected, got, topStackFiles: stackFiles(out, PY_FILE_LINE).slice(0, 5) });
    }
    const first = summary.failures[0];
    summary.nextAction = first?.topStackFiles[0]
      ? `Localize the failure in ${first.topStackFiles[0]} and fix the smallest cause of "${first.message || first.test}".`
      : "Localize and fix the failing test's root cause without editing the test.";
    return summary;
  }

  // 5) jest / vitest failures
  if (/(●|✕|FAIL )\s/.test(out) || /Expected:/.test(out)) {
    summary.isTestFailure = true;
    const names = uniq([...out.matchAll(/^\s*(?:●|✕)\s+(.+?)\s*$/gm)].map((m) => m[1].trim())).slice(0, 6);
    const expected = out.match(/Expected:?\s*(.+)/)?.[1]?.trim().slice(0, 200) ?? "";
    const got = out.match(/Received:?\s*(.+)/)?.[1]?.trim().slice(0, 200) ?? "";
    const files = stackFiles(out, JS_STACK_FILE).filter((f) => !f.includes("node_modules"));
    summary.summary = names.length ? `${names.length} test(s) failed` : "jest/vitest failure";
    for (const name of (names.length ? names : ["(test)"])) {
      summary.failures.push({ test: name, errorType: "AssertionError", message: (expected || got) ? `expected ${expected} / received ${got}` : "", expected, got, topStackFiles: files.slice(0, 5) });
    }
    summary.nextAction = files[0]
      ? `Fix ${files[0]} so it produces ${expected || "the expected value"} instead of ${got || "the received value"}.`
      : "Fix the code under test to match the expected value.";
    return summary;
  }

  // 6) Generic runtime error
  const err = out.match(GENERIC_ERR);
  if (err) {
    summary.errorType = err[1];
    const msg = out.match(new RegExp(`${err[1]}:\\s*(.+)`))?.[1]?.trim().slice(0, 200);
    summary.summary = msg ? `${err[1]}: ${msg}` : `${err[1]} raised`;
    summary.failures.push({ test: "(runtime)", errorType: err[1], message: summary.summary, expected: "", got: "", topStackFiles: uniq([...stackFiles(out, PY_FILE_LINE), ...stackFiles(out, JS_STACK_FILE)]).slice(0, 5) });
    summary.nextAction = `Reproduce and fix the ${err[1]} at the top stack frame.`;
    return summary;
  }

  summary.summary = `command exited ${exitCode}`;
  summary.nextAction = "Inspect the tail output below and fix the reported failure.";
  return summary;
}

function pytestFailureBody(out: string, testName: string): string {
  const short = testName.includes("::") ? testName.split("::").pop()! : testName;
  const esc = short.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const start = out.match(new RegExp(`^_{3,}\\s+${esc}\\s+_*\\s*$`, "m")) ?? out.match(new RegExp(`^FAILED\\s+${esc}.*$`, "m"));
  if (!start || start.index === undefined) return out;
  const from = start.index + start[0].length;
  const rest = out.slice(from);
  const nextSep = rest.match(/^(?:_{3,}|=+\s+\d)/m);
  return nextSep && nextSep.index !== undefined ? rest.slice(0, nextSep.index) : rest;
}

function extractExpectedGot(block: string): [string, string] {
  const assertEq = block.match(/assert\s+(.+?)\s*==\s*(.+?)$/m);
  if (assertEq) return [assertEq[2].trim().slice(0, 200), assertEq[1].trim().slice(0, 200)];
  const expected = block.match(/Expected:?\s*(.+)/)?.[1]?.trim().slice(0, 200) ?? "";
  const got = block.match(/(?:Received|Actual|got):?\s*(.+)/i)?.[1]?.trim().slice(0, 200) ?? "";
  return [expected, got];
}

/** Render a compact card for the model, always ending with one imperative next-action line. */
export function renderFailureCard(summary: FailureSummary, maxChars = 900): string {
  const out: string[] = ["=== FAILURE SUMMARY ==="];
  out.push(`command: ${summary.command}`);
  out.push(`exit: ${summary.exitCode} | ${summary.summary}`);
  const flags = [summary.isSyntaxError && "SYNTAX", summary.isImportError && "IMPORT", summary.isTestFailure && "TEST"].filter(Boolean);
  if (flags.length) out.push(`flags: ${flags.join(" ")}`);
  for (const [i, f] of summary.failures.slice(0, 4).entries()) {
    out.push(`[${i + 1}] ${f.test}${f.errorType ? ` (${f.errorType})` : ""}`);
    if (f.message) out.push(`    ${f.message}`);
    if (f.expected || f.got) out.push(`    expected: ${f.expected} | got: ${f.got}`);
    if (f.topStackFiles.length) out.push(`    files: ${f.topStackFiles.slice(0, 3).join(", ")}`);
  }
  out.push(`NEXT: ${summary.nextAction}`);
  return clip(out.join("\n"), maxChars);
}

/** Convenience: raw output -> compact model-facing card in one call. */
export function compressFailureOutput(command: string, stdout: string, stderr: string, exitCode: number, maxChars = 900): string {
  return renderFailureCard(compressCommandResult(command, stdout, stderr, exitCode), maxChars);
}

File: cheater/failure_compressor.py
"""Cheater failure compressor: turn raw pytest / py_compile / runtime output
into a tight, model-friendly JSON summary.

Why this exists:
  - A raw pytest log is hundreds or thousands of lines. A 9B model chokes on
    that. We need a single small JSON blob that points the model at the
    right test, the right error type, and the right files.
  - This module is deterministic. The LLM does NOT see raw output, ever.

Public API:
  compress_command_result(command, stdout, stderr, exit_code)
        -> FailureSummary
  render_failure_for_model(summary, max_tokens=3000) -> str

The FailureSummary dataclass serializes to:
  {
    "command": "...",
    "passed": bool,
    "exit_code": int,
    "failures": [
       {"test": ..., "error_type": ..., "message": ...,
        "expected": ..., "got": ...,
        "top_stack_files": [...], "likely_relevant_files": [...]}
    ],
    "summary": "...",
    "is_syntax_error": bool,
    "is_import_error": bool,
    "is_test_failure": bool
  }
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any

# pytest patterns
_PYTEST_FAILED_LINE_RE = re.compile(r"^FAILED\s+([\w./:\\-]+?)(?:\s*-\s*|\s*$)", re.MULTILINE)
_PYTEST_PASSED_RE = re.compile(r"=+\s*(\d+)\s+passed")
_PYTEST_FAILED_RE = re.compile(r"=+\s*(\d+)\s+failed")
# Strip ANSI / trailing =s inside the captured group
_PYTEST_SHORT_SUMMARY_RE = re.compile(
    r"=+\s*((?:\d+\s+)?(?:passed|failed|error|skipped|warning)[^\n=]*?)\s*=+"
)
_PYTEST_TEST_LINE_RE = re.compile(r"^(FAILED\s+\S+)$", re.MULTILINE)
_PY_ASSERT_RE = re.compile(r"AssertionError:?\s*(.*)", re.IGNORECASE)
_PY_ASSERT_DETAIL_RE = re.compile(r"assert\s+(.+?)\s*$", re.MULTILINE)
_PY_FILE_LINE_RE = re.compile(r"""File ['"]([^'"]+)['"], line (\d+)""")
_PY_SYNTAX_HINT_RE = re.compile(
    r"File ['\"]([^'\"]+)['\"], line (\d+).*?(?:SyntaxError|IndentationError|TabError)",
    re.DOTALL,
)
_PY_IMPORT_ERROR_RE = re.compile(r"(?:^|\n)(?:from\s+([\w.]+)\s+import\s+([\w, ]+)|import\s+([\w.]+))")
_PY_MOD_NOT_FOUND_RE = re.compile(r"ModuleNotFoundError: No module named '([^']+)'")
_PY_GENERIC_ERR_RE = re.compile(r"\b(\w+(?:Error|Exception|Interrupt|Exit|Fatal))\b")

_TAIL_KEEP = 1500
_TAIL_MAX_LINES = 80


# ---------- data types ----------

@dataclass
class FailureEntry:
    test: str = ""
    error_type: str = ""
    message: str = ""
    expected: str = ""
    got: str = ""
    top_stack_files: list[str] = field(default_factory=list)
    likely_relevant_files: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class FailureSummary:
    command: str = ""
    passed: bool = True
    exit_code: int = 0
    failures: list[FailureEntry] = field(default_factory=list)
    summary: str = ""
    is_syntax_error: bool = False
    is_import_error: bool = False
    is_test_failure: bool = False
    tail: str = ""
    error_type: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "command": self.command,
            "passed": self.passed,
            "exit_code": self.exit_code,
            "failures": [f.to_dict() for f in self.failures],
            "summary": self.summary,
            "is_syntax_error": self.is_syntax_error,
            "is_import_error": self.is_import_error,
            "is_test_failure": self.is_test_failure,
            "error_type": self.error_type,
        }


# ---------- helpers ----------

def _clip(text: str, max_chars: int) -> str:
    if not text or len(text) <= max_chars:
        return text or ""
    return text[:max(0, max_chars - 30)] + "\n... [truncated]"


def _tail(text: str, max_lines: int = _TAIL_MAX_LINES, max_chars: int = _TAIL_KEEP) -> str:
    if not text:
        return ""
    lines = text.splitlines()
    tail = "\n".join(lines[-max_lines:])
    return _clip(tail, max_chars)


def _stack_files(output: str) -> list[str]:
    """Extract unique file paths from a Python traceback (top first)."""
    seen: list[str] = []
    seen_set: set[str] = set()
    for m in _PY_FILE_LINE_RE.finditer(output):
        path = m.group(1).replace("\\", "/")
        if path not in seen_set:
            seen.append(path)
            seen_set.add(path)
    return seen


def _strip_ansi(text: str) -> str:
    return re.sub(r"\x1b\[[0-9;]*[a-zA-Z]", "", text or "")


# ---------- main entry point ----------

def compress_command_result(
    command: str,
    stdout: str,
    stderr: str,
    exit_code: int,
) -> FailureSummary:
    """Parse the output and return a FailureSummary. Never raises."""
    s = FailureSummary(command=command, exit_code=exit_code)
    out = _strip_ansi((stdout or "") + "\n" + (stderr or ""))
    s.tail = _tail(out)
    s.passed = exit_code == 0
    if s.passed:
        s.summary = "command exited 0"
        return s
    # 1) Syntax error
    sm = _PY_SYNTAX_HINT_RE.search(out)
    if sm:
        s.is_syntax_error = True
        # Detect error type from the matched range (last 20 chars)
        tail = sm.group(0).split("\n")[-1].strip()
        for et in ("SyntaxError", "IndentationError", "TabError"):
            if et in tail:
                s.error_type = et
                break
        if not s.error_type:
            s.error_type = "SyntaxError"
        s.summary = f"Python {s.error_type} at {sm.group(1)}:L{sm.group(2)}"
        s.failures.append(FailureEntry(
            test="(no test)",
            error_type=s.error_type,
            message=s.summary,
            top_stack_files=[sm.group(1).replace("\\", "/")],
            likely_relevant_files=[sm.group(1).replace("\\", "/")],
        ))
        return s
    # 2) Import error
    if "ModuleNotFoundError" in out or "ImportError" in out:
        s.is_import_error = True
        m = _PY_MOD_NOT_FOUND_RE.search(out)
        if m:
            s.error_type = "ModuleNotFoundError"
            s.summary = f"ModuleNotFoundError: {m.group(1)}"
        else:
            m2 = _PY_GENERIC_ERR_RE.search(out)
            s.error_type = m2.group(1) if m2 else "ImportError"
            s.summary = "import failed"
        # Try to grab the offending line
        for imp in _PY_IMPORT_ERROR_RE.finditer(out)[:3] if False else _PY_IMPORT_ERROR_RE.finditer(out):
            pass
        # Find the traceback top
        stack = _stack_files(out)
        s.failures.append(FailureEntry(
            test="(import)",
            error_type=s.error_type,
            message=s.summary,
            top_stack_files=stack[:5],
            likely_relevant_files=stack[:3],
        ))
        return s
    # 3) Pytest failure
    failed_tests = list(dict.fromkeys(m.group(1) for m in _PYTEST_FAILED_LINE_RE.finditer(out)))
    if failed_tests or "FAILED" in out or "= " in out and " failed " in out:
        s.is_test_failure = True
        # Summary line
        pm = _PYTEST_SHORT_SUMMARY_RE.search(out)
        if pm:
            s.summary = pm.group(1).strip()
        else:
            s.summary = f"{len(failed_tests)} test(s) failed"
        # Pull each failure block. Pytest writes a '_____<test>_____' separator
        # before the test body, and the body ends at the next separator
        # ('_____' or '====' final summary).
        for test_name in failed_tests[:10]:
            short = test_name.split("::")[-1] if "::" in test_name else Path(test_name).stem
            esc = re.escape(short)
            # Try a few separator shapes
            body_start = None
            for pat in (
                rf"^_{5,}\s+{esc}\s+_{5,}\s*$",
                rf"^_{5,}\s+{esc}\s*$",
                rf"^_{5,}\s+{re.escape(test_name)}\s+_{5,}\s*$",
            ):
                bm = re.search(pat, out, re.MULTILINE)
                if bm:
                    body_start = bm.end()
                    break
            if body_start is None:
                # Fall back to the FAILED line
                tm = re.search(r"^FAILED\s+" + re.escape(test_name) + r".*$",
                               out, re.MULTILINE)
                if not tm:
                    continue
                body_start = tm.end()
            # End: next '_____' or '====' final summary
            next_sep = re.search(r"^(?:_{5,}|=+\s+\d)", out[body_start:], re.MULTILINE)
            body_end = body_start + next_sep.start() if next_sep else len(out)
            body = out[body_start:body_end]
            e = FailureEntry(test=test_name)
            e.top_stack_files = _stack_files(body)
            if not e.top_stack_files and "::" in test_name:
                # The body didn't include a stack; use the test file path
                tf = test_name.split("::", 1)[0]
                e.top_stack_files = [tf]
            e.likely_relevant_files = e.top_stack_files[:3]
            et = _PY_GENERIC_ERR_RE.search(body)
            if et:
                e.error_type = et.group(1)
            am = _PY_ASSERT_RE.search(body)
            if am:
                e.message = am.group(1).strip()[:300]
            if not e.message:
                for ln in body.splitlines():
                    ls = ln.strip()
                    if ls.startswith("E   ") or ls.startswith("E "):
                        e.message = ls.lstrip("E ").strip()[:300]
                        break
            if not e.message:
                for ln in body.splitlines():
                    ls = ln.strip()
                    if ls and not ls.startswith(">"):
                        e.message = ls[:300]
                        break
            e.expected, e.got = _extract_expected_got(body)
            s.failures.append(e)
        # Fallback if we still built nothing
        if not s.failures and failed_tests:
            for t in failed_tests[:10]:
                e = FailureEntry(test=t)
                e.top_stack_files = _stack_files(out)
                e.likely_relevant_files = e.top_stack_files[:3]
                s.failures.append(e)
        return s
    # 4) Generic Python runtime error
    m = _PY_GENERIC_ERR_RE.search(out)
    if m:
        s.error_type = m.group(1)
        # Try to extract the error message on the same line (e.g. "ValueError: oh no")
        line_re = re.compile(rf"^{re.escape(s.error_type)}:\s*(.+?)$", re.MULTILINE)
        lm = line_re.search(out)
        if lm:
            s.summary = f"{s.error_type}: {lm.group(1).strip()[:200]}"
        else:
            s.summary = f"{s.error_type} raised"
    else:
        s.summary = f"command exited {exit_code}"
    s.failures.append(FailureEntry(
        test="(runtime)",
        error_type=s.error_type or "RuntimeError",
        message=s.summary,
        top_stack_files=_stack_files(out)[:5],
        likely_relevant_files=_stack_files(out)[:3],
    ))
    return s


# ---------- expected/got extraction ----------

def _extract_expected_got(block: str) -> tuple[str, str]:
    """Best-effort parse of 'assert X == Y' or pytest AssertionError."""
    expected = ""
    got = ""
    # pytest "AssertionError: <a> == <b>" or "assert <a> == <b>"
    m = re.search(r"assert\s+(.+?)\s*==\s*(.+?)$", block, re.MULTILINE)
    if m:
        got = m.group(1).strip()[:200]
        expected = m.group(2).strip()[:200]
    if not expected:
        m = re.search(r"AssertionError:\s*(.+?)\s*==\s*(.+)", block)
        if m:
            got = m.group(1).strip()[:200]
            expected = m.group(2).strip()[:200]
    return expected, got


# ---------- rendering ----------

def render_failure_for_model(summary: FailureSummary, max_tokens: int = 3000) -> str:
    """Render the summary as a tight markdown block for the model."""
    cap = max(400, max_tokens * 4)
    out: list[str] = []
    out.append("=== TEST FAILURE SUMMARY ===")
    out.append(f"command:    {summary.command}")
    out.append(f"exit_code:  {summary.exit_code}")
    out.append(f"passed:     {summary.passed}")
    if summary.error_type:
        out.append(f"error_type: {summary.error_type}")
    out.append(f"summary:    {summary.summary}")
    flags: list[str] = []
    if summary.is_syntax_error:
        flags.append("SYNTAX_ERROR")
    if summary.is_import_error:
        flags.append("IMPORT_ERROR")
    if summary.is_test_failure:
        flags.append("TEST_FAILURE")
    if flags:
        out.append(f"flags:      {' '.join(flags)}")
    if summary.failures:
        out.append("")
        out.append(f"failures ({len(summary.failures)}):")
        for i, f in enumerate(summary.failures[:10], start=1):
            out.append(f"  [{i}] test: {f.test or '(unknown)'}")
            if f.error_type:
                out.append(f"      error_type: {f.error_type}")
            if f.message:
                out.append(f"      message:    {f.message}")
            if f.expected or f.got:
                out.append(f"      expected:   {f.expected}")
                out.append(f"      got:        {f.got}")
            if f.top_stack_files:
                out.append(f"      stack:      {', '.join(f.top_stack_files[:5])}")
            if f.likely_relevant_files:
                out.append(f"      files:      {', '.join(f.likely_relevant_files[:3])}")
    if summary.tail:
        out.append("")
        out.append("tail (last lines, raw):")
        out.append("```")
        out.append(_clip(summary.tail, max(200, cap // 2)))
        out.append("```")
    out.append("")
    out.append("HINTS:")
    out.append("- Do NOT edit tests unless explicitly allowed.")
    out.append("- Use localize_bug to find the suspect source files.")
    out.append("- Use inspect_symbol to see callers/definitions.")
    out.append("- Use read_context_pack for a compact context block.")
    text = "\n".join(out)
    return _clip(text, cap)
